### PR TITLE
Support RaRef and ResRef

### PR DESCRIPTION
**Pull request by Roms1383**
This PR adds support for RaRef and ResRef, along with tests.

### CODE REVIEW

File: src/types/res.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+impl Clone for ResRef {
+    fn clone(&self) -> Self {
+        Self(red::ResRef {
+            resource: self.0.resource,
+        })
+    }
+}
+
+fn encode_path(path: &(impl AsRef<Path> + ?Sized)) -> Result<u64, ResourcePathError> {
```
Comment (jac3km4) [inline src/types/res.rs:48]:
```suggestion
fn encode_path(path: impl AsRef<Path>) -> Result<u64, ResourcePathError> {
```

File: src/types/res.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+use crate::raw::root::RED4ext as red;
+
+pub const MAX_LENGTH: usize = 216;
+
+#[derive(Debug, Default, Clone, Copy)]
+#[repr(transparent)]
+pub struct RaRef(red::RaRef);
+
+impl RaRef {
+    pub fn new(path: &(impl AsRef<Path> + ?Sized)) -> Result<Self, ResourcePathError> {
```
Comment (jac3km4) [inline src/types/res.rs:15]:
```suggestion
    pub fn new(path: impl AsRef<Path>) -> Result<Self, ResourcePathError> {
```

File: src/types/res.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    };
+    ($base:expr, ) => {
+        $base
+    };
+    ($lit:literal $($tt:tt)*) => {
+        $crate::types::ResRef::new(
+            &$crate::res_ref!($lit, $($tt)*).to_string()
+        )
+    };
+}
```
Comment (jac3km4) [inline src/types/res.rs:105]: this is calling `<[]>::join` which seems to be an incorrect way to implement this macro (the test doesn't check the created paths),
I'd fall back to the implementation in red4ext-rs
```suggestion
#[macro_export]
macro_rules! res_ref {
    ($base:expr, /$lit:literal $($tt:tt)*) => {
        $crate::res_ref!($base.join($lit), $($tt)*)
    };
    ($base:expr, ) => {
        $base
    };
    ($lit:literal $($tt:tt)*) => {
        $crate::ResRef::new(
            &$crate::res_ref!(::std::path::Path::new($lit), $($tt)*).to_string_lossy()
        )
    };
}
```

File: src/types/res.rs (hunk 1 above)
Comment (jac3km4) [reply src/types/res.rs:48]: passing something convertible to a reference behind a reference seems counterproductive

Comment (Roms1383) [reply src/types/res.rs]: fixed in b00bb6f

File: src/types/res.rs (hunk 2 above)
Comment (Roms1383) [reply src/types/res.rs]: fixed in b00bb6f

File: src/types/res.rs (hunk 3 above)
Comment (Roms1383) [reply src/types/res.rs]: adjusted in c9f1164, also completed test for created paths.

**Review by jac3km4 — APPROVED** [review]

**Pull request merged**